test/test_factbase.rb

```diff
@@ -219,4 +220,164 @@ def test_txn_with_rollback
+      n += 1
+    end
+    assert_equal(100, fb.size)
+    i = 0
+    Threads.new(100) do
+      f = fb.query("(eq foo #{i})").each.to_a
+      assert_equal(1, f.count)
+      assert_equal(i, f.first.foo)
+      i += 1
+    end
+  end
+
+  def test_different_properties_when_concurrent_inserts
+    fb = Factbase.new
+    n = 1
+    Threads.new(100).assert do
+      fb.insert.send(:"prop_#{n}=", n)
+      n += 1
+    end
+    assert_equal(100, fb.size)
+    100.times do |num|
+      i = num + 1
+      prop = "prop_#{i}"
+      f = fb.query("(eq #{prop} #{i})").each.to_a
+      assert_equal(1, f.count)
+      assert_equal(i, f.first.send(prop.to_sym))
+    end
+  end
+
+  # sometimes this test fails
+  def test_concurrent_transactions_inserts
+    fb = Factbase.new
+    i = 0
+    Threads.new(100).assert do
+      fb.txn do |fbt|
+        fact = fbt.insert
+        fact.thread_id = i
+        i += 1
+      end
+    end
+    assert_equal(100, fb.size)
+    assert_equal(100, fb.query('(exists thread_id)').each.to_a.size)
+  end
+
+  def test_concurrent_transactions_with_rollbacks
+    fb = Factbase.new
+    Threads.new(100).assert do |i|
+      fb.txn do |fbt|
+        fact = fbt.insert
+        fact.thread_id = i
+        raise Factbase::Rollback
+      end
+    end
+    assert_equal(0, fb.size)
+  end
+
+  # sometimes this test fails
+  def test_concurrent_transactions_successful
+    fb = Factbase.new
+    i = 0
+    Threads.new(100).assert do
+      fb.txn do |fbt|
+        fact = fbt.insert
+        fact.thread_id = i
+        fact.value = i * 10
+        i += 1
+      end
+    end
+    facts = fb.query('(exists thread_id)').each.to_a
+    assert_equal(100, facts.size)
+    facts.each do |fact|
+      assert_equal(fact.value, fact.thread_id * 10)
+    end
+  end
+
+  def test_concurrent_queries
+    fb = Factbase.new
+    i = 0
+    Threads.new(100).assert do
+      fact = fb.insert
+      fact.thread_id = i
+      fact.value = i * 10
+      i += 1
+    end
+    Threads.new(100).assert do
+      results = fb.query('(exists thread_id)').each.to_a
+      assert_equal(100, results.size)
+
+      thread_ids = results.map(&:thread_id)
+      assert_equal((0..99).to_a, thread_ids.sort)
+    end
+  end
+
+  def test_export_import_concurrent
+    fb = Factbase.new
+    mutex = Mutex.new
+    Threads.new(100).assert do
+      fact = fb.insert
+      fact.value = 42
+    end
+    exported = []
+    Threads.new(5).assert do
+      mutex.synchronize do
+        exported << fb.export
+      end
+    end
+    assert_equal(5, exported.size)
+    i = 0
+    Threads.new(5).assert do
+      new_fb = Factbase.new
+      new_fb.import(exported[i])
+      assert_equal(fb.size, new_fb.size)
+      facts = fb.query('(eq value 42)').each.to_a
+      assert_equal(100, facts.size)
+      facts.each do |fact|
+        new_fact = new_fb.query("(eq value #{fact.value})").each.to_a.first
+        assert_equal(fact.value, new_fact.value)
+      end
+      i += 1
```

Comment:
It's not clear why you increment the variable here and don't check it later. If all the threads start simultaneously, then `exported[i]` will be `exported[0]`

Also, it's worth keeping in mind that `i += 1` is not an atomic operation. It's actually two operations

```ruby
temp = i + 1
i = temp
```

And there is a race condition. I mean, if you will check `i` after the block, it will most likely be less than 5. And some of the threads used the same `fb.` I don't think it's expected behavior.

If you want each thread to have its own `fb` you should use the argument that `threads` pass into the block